virtual_rainforest/core/core_components.py

```diff
@@ -0,0 +1,294 @@
+from virtual_rainforest.core.logger import LOGGER
+
+
+@dataclass
+class CoreComponents:
+    """Core model components.
+
+    This dataclass takes a validated model configuration and uses it to generate a set
+    of core model attributes, populated via the ``__init__`` method of
+    :class:`~virtual_rainforest.core.base_model.BaseModel` and hence inherited by the
+    specific model subclasses.
+    """
+
+    layer_structure: LayerStructure = field(init=False)
+    """The vertical layer structure for the simulation."""
+    model_timing: ModelTiming = field(init=False)
+    """The model timing details for the simulation."""
+    core_constants: ConstantsDataclass = field(init=False)
+    """The core constants definitions for the simulation"""
+    config: InitVar[Config]
+    """A validated model configuration."""
+
+    def __post_init__(self, config: Config) -> None:
+        """Populate the core components from the config."""
+        self.layer_structure = LayerStructure(config=config)
+        self.model_timing = ModelTiming(config=config)
+        self.core_constants = load_constants(config, "core", "CoreConsts")
+
+
+@dataclass
+class ModelTiming:
+    """Model timing details.
+
+    This data class defines the timing of a Virtual Rainforest simulation from the
+    ``core.timing`` section of a validated model configuration. The start time, run
+    length and update interval are all extracted from the configuration and validated.
+
+    The end time is calculated from the previously extracted timing information. This
+    end time will always be the largest whole multiple of the update interval that
+    exceeds or equal the configured ``run_length``.
+
+
+    Raises:
+        ConfigurationError: If the timing configuration details are incorrect.
+    """
+
+    start_time: np.datetime64 = field(init=False)
+    """The start time of the simulation."""
+    end_time: np.datetime64 = field(init=False)
+    """The calculated end time of the simulation."""
+    reconciled_run_length: np.timedelta64 = field(init=False)
+    """The difference between start and calculated end time."""
+    run_length: np.timedelta64 = field(init=False)
+    """The configured run length."""
+    run_length_quantity: Quantity = field(init=False)
+    """The configured run length as a pint Quantity."""
+    update_interval: np.timedelta64 = field(init=False)
+    """The configured update interval."""
+    update_interval_quantity: Quantity = field(init=False)
+    """The configured update interval as a pint Quantity."""
+    config: InitVar[Config]
+    """A validated model configuration."""
+
+    def __post_init__(self, config: Config) -> None:
+        """Populate the ``ModelTiming`` instance.
+
+        This method populates the ``ModelTiming`` attributes from the provided
+        :class:`~virtual_rainforest.core.config.Config` instance.
+
+        Args:
+            config: A Config instance.
+        """
+
+        timing = config["core"]["timing"]
```

Comment:
By the time this is called, the config has already validated to ensure there's a `core` and a `timing`, right? I think so, but just cross checking.